specfem::Context struct  and ContextGuard for initialization of kokkos, mpi, and handling initialization from python side etc.

lsawade opened the pull request:
## Description

When I was trying out the Kokkos::View extents syntax, I asked claude to create a basic cpp project for me to just play around with the dependencies, and not having to wrestle with specfem compilation.

It ended up creating a core initialization framework to handle initialization and ending of Kokkos.

I created this PR that implements everything. Everything is tested include the Python interface, so no issues there.

<details>
<summary> Claude's sample implementation (Click me)</summary>

```cpp

namespace specfem {

/**
 * @brief Minimal SPECFEM++ core class for testing dependencies
 * 
 * This class provides a simple interface to test that all major
 * dependencies (Kokkos, YAML-CPP, Boost) are working correctly.
 */
class Core {
public:
    Core();
    ~Core() {if (initialized) finalize()};
    void initialize(int argc, char* argv[]);
    void finalize();
    void print_system_info();
    void run_all_tests(argc, argv);
private:
    bool kokkos_initialized_;
    std::string execution_space_;
};
```
`main` function that uses it
```cpp
int main(int argc, char* argv[]) {
    specfem::Core core;
    
    try {
        core.initialize(argc, argv);
        bool success = core.run_all_tests(argc, argv);

        return success ? 0 : 1;
        
    } catch (const std::exception& e) {
        std::cerr << "Fatal error: " << e.what() << std::endl;
        return 1;
    }
}
```
</details>

What's interesting is that it a allows for a pretty clean single executable without a lot of multiplied code:

running 2d:
```bash
specfem 2d -p specfem_config.yaml
```

running 3d:
```bash
specfem 3d -p specfem_config.yaml
```

Let me know what you think.

## Issue Number

If there is an issue created for these changes, link it here

## Checklist

Please make sure to check developer documentation on specfem docs.

- [ ] I ran the code through pre-commit to check style
- [ ] **THE DOCUMENTATION BUILDS WITHOUT WARNINGS/ERRORS**
- [ ] I have added labels to the PR (see right hand side of the PR page)
- [ ] My code passes all the integration tests
- [ ] I have added sufficient unittests to test my changes
- [ ] I have added/updated documentation for the changes I am proposing
- [ ] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

lsawade committed:
Creates files that take care of initialization and state keeping

lsawade committed:
Removed inverted dependency, dont know why that is building locally..

lsawade commented:
If approved, this will need additional file edits for the documentation.

Rohit-Kakodkar reviewed (changes requested):
I am mostly on board with this. 

- I think we need a new enumeration that is different from dimension. How about
```
namespace specfem {
enum class execution_model {
dim2, dim3, globe, ...
}
}
```
- Use `std::unique_ptr` where you can if the ptr is owned exclusively by core (which I think is the case). Or use `std::shared_ptr` with `weak_ptr` which might be appropriate as well. I recomment seeing Cherno's videos on those if you haven't.

lsawade commented:
> I am mostly on board with this.

Ok! I'm going to wait for @icui 's review as well before making the requested updates

> Use `std::unique_ptr` where you can if the ptr is owned exclusively by core (which I think is the case). Or use `std::shared_ptr` with `weak_ptr` which might be appropriate as well. I recomment seeing Cherno's videos on those if you haven't.

I don't think I watched that one yet, will do!

icui commented:
Looks good to me. Shall we rename it to `Context` since `Core` is already a folder that includes everything?

Rohit-Kakodkar commented on core/specfem/core.cpp:
Look at `Kokkos::ScopedGuard` it implements the context functionality for Kokkos

lsawade replied:
I think that would break the python interface. I created a `specfem::ContextGuard` instead that does the same thing.

lsawade committed:
Merge branch 'devel' into specfem-core-framework

lsawade committed:
Addressed all the comments from the PR.

lsawade commented:
I still have to update the documentation. But I thought since the updates were larger, I still wait for it.

lsawade committed:
Missing include

lsawade committed:
Removed static instance, removed kokkos init and finalize. ContextGuard fully manages the created context. For Python we still create a static instance that is checked against. This should theoretically allow for multiple Kokkoses in one executable.

Rohit-Kakodkar commented on core/specfem/execute.hpp:
```diff
@@ -0,0 +1,47 @@
+#include <yaml-cpp/yaml.h>
+
+// Legacy execute function (currently used by existing code)
+void execute(
```
Missing namespace for these functions.

Rohit-Kakodkar commented on core/specfem/simulation/context.hpp:
```diff
@@ -0,0 +1,162 @@
+ * dimension-templated execution, and proper resource cleanup.
+ * Typically used through ContextGuard for scoped lifetime management.
+ */
+class Context {
```
Can we have a 3 level namespace for `context` and `execute`

lsawade replied:
specfem::context::Context() and specfem::execute::execute()?

Rohit-Kakodkar commented on include/enumerations/execution.hpp:
```diff
@@ -0,0 +1,60 @@
+ * Currently supports 2D and 3D simulations, with room for future
+ * expansion (e.g., globe simulations).
+ */
+enum class model { dim2, dim3 };
```
Should we name these `Cartesian2D`, `Cartesian3D`, `Axisem2D`, `Globe3D` etc.

Rohit-Kakodkar commented on src/python/core.cpp:
```diff
@@ -1,80 +1,71 @@
-    std::strcpy(argv[i], str.c_str()); // Copy the string content
+  if (global_context_guard != nullptr) {
+    return false; // Already initialized
   }
```
You can directly evaluate a `shared_ptr`, i.e. `if (global_context_guard)` is same as `if (global_context_guard != nullptr)`

Rohit-Kakodkar commented on CMakeLists.txt:
Can you update the configure tests variable too. I have it prefixed with `SPECFEMPP`

lsawade committed:
Updates to namespaces, added specfem3d to context and execution

- put context and execution in deeper namespaces
- updated the original specfem2d and specfem3d executables to
  use ContextGuard. They are kept for compatibility for now.
- add print functions to mpi object

lsawade committed:
Updated the header print function.

lsawade committed:
Moved specfem::execute::... to specfem::simulation::...

lsawade committed:
Renames execution::model -> simulation::model

- Updates names of the models
- deprecates enumerations::execution
- adds content to enumerations::simulation
- updates references

lsawade committed:
Remove the nullptr comparison

Rohit-Kakodkar commented on core/specfem/context.cpp:
```diff
@@ -0,0 +1,129 @@
+  }
+}
+
+bool Context::execute_with_dimension(
```
Can you not just call this `execute` as well. The argument list should distinguish between these 2 functions.

Rohit-Kakodkar commented on core/specfem/context.hpp:
```diff
@@ -0,0 +1,161 @@
+   */
+  template <specfem::dimension::type DimensionTag>
+  bool
+  execute(const YAML::Node &parameter_dict, const YAML::Node &default_dict,
```
I have my doubts on if these functions should be here. In my view a context should only be setting up the context for `Kokkos` and `MPI`. Context shouldnt be running the simulation. 

```cpp

int main (...) {
    specfem::simulation::ContextGuard context(...);
    specfem::simulation::execute(...)
}

```

Would the above not work?

lsawade replied:
true, let me try.

Rohit-Kakodkar commented on core/specfem/simulation.hpp:
```diff
@@ -0,0 +1,57 @@
+void simulation_3d(
+    const YAML::Node &parameter_dict, const YAML::Node &default_dict,
+    std::vector<std::shared_ptr<specfem::periodic_tasks::periodic_task> > tasks,
+    specfem::MPI::MPI *mpi);
```
These functions are internal right? I think we shouldnt expose them.

Rohit-Kakodkar commented on core/specfem/simulation.hpp:
```diff
@@ -0,0 +1,57 @@
+// refactored
+
+template <specfem::dimension::type DIM>
+void execute_templated(
```
```suggestion
void execute(
```

Rohit-Kakodkar commented on core/specfem/simulation/context.hpp:
```diff
@@ -0,0 +1,161 @@
+ * dimension-templated execution, and proper resource cleanup.
+ * Typically used through ContextGuard for scoped lifetime management.
+ */
+class Context {
```
Context should also go under `specfem::simulation` namespace

Rohit-Kakodkar commented on core/specfem/CMakeLists.txt:
```diff
@@ -6,3 +7,49 @@ add_subdirectory(source)
+
+
+# Core library - context manager for initialization/finalization
+add_library(specfem_core STATIC context.cpp)
```
You need another subdirectory called simulation.

lsawade replied:
this was a remnant

lsawade committed:
Disentangle execute from context

- specfem::simulation now contains both context and execute function
- the execute function takes as an argument the 2d and 3d
- calls can be made like so
```cpp
try {
  context = specfem::simulation::contextguard()
  success = specfem::simulation::execute("2d", ...)
```

Rohit-Kakodkar commented on core/specfem/simulation/context.hpp:
```diff
@@ -0,0 +1,131 @@
+#include <vector>
+#include <yaml-cpp/yaml.h>
+
+namespace specfem {
```
```suggestion
namespace specfem::simulation {
```

lsawade committed:
Renamed simulation to program and moved context namespace.

lsawade committed:
Add new static struct specfem::MPI that holds rank, size, and comm information.

Currently it is named MPI_new, until we deprecated the old mpi struct

lsawade committed:
Forgot to add the header file

lsawade committed:
Deprecated the context - context guard combination in favor of a single context struct

lsawade committed:
Added the logger and the logger options structs

lsawade committed:
This is unfortunately doing a lot.

- This converts mpi->cout() to the appropriate specfem::Logger::<loglevel>()
- Since some of these function are being used by the tests, I needed to
  deprecate {Kokkos,MPI}_environment in favor of SPECFEM_Environement
- So alot of the tests were updated for that reason.

lsawade committed:
This deprecates mpi->cout in favor of Logging methods.

The next step will be removing the specfem::MPI::MPI object and function passes.

lsawade committed:
Merge branch 'devel' into specfem-core-framework

lsawade committed:
Merge branch 'specfem-core-framework' into specfem-core-framework-depr-mpi-cout

lsawade commented:
I merged devel into #1416 by accident, so we just gotta wait for that one to merge into here, for the conflicts to be resolved.

lsawade committed:
Merge branch 'devel' into specfem-core-framework-depr-mpi-cout

lsawade committed:
Updated the file extension, and includes for the legacy executables

lsawade committed:
Added MPI abort functionality with line indication.

lsawade committed:
Merge branch 'specfem-core-framework-logger-impl' into specfem-core-framework-depr-mpi-cout

Rohit-Kakodkar committed:
Merge pull request #1416 from PrincetonUniversity/specfem-core-framework-depr-mpi-cout

Specfem Core Framework Depr Mpi Cout

lsawade committed:
Implemented check_context for mpi abotr

lsawade committed:
Merge branch 'specfem-core-framework-logger-impl' of github.com:PrincetonUniversity/SPECFEMPP into specfem-core-framework-logger-impl

lsawade committed:
Merge pull request #1415 from PrincetonUniversity/specfem-core-framework-logger-impl

Specfem-Core-Framework-Logger-Impl

lsawade committed:
Merge branch 'devel' into specfem-core-framework

lsawade committed:
Merge branch 'specfem-core-framework' of github.com:PrincetonUniversity/SPECFEMPP into specfem-core-framework
